ath79: add support for TP-Link TL-WR841HP v3

teqlee opened the pull request:
Specifications:
- QCA9533 SoC, 8 MB nor flash, 64 MB DDR2 RAM
- 4x Ethernet LAN 10/100, 1x Ethernet WAN 10/100
- 1x WAN, LAN, Wifi, PWR, WPS, RE Leds
- Reset, Wifi on/off, WPS, RE buttons
- Serial UART at J4 onboard: 3.3v GND RX TX, 1152008N1

Label MAC addresses based on vendor firmware, example:
LAN    *:ea    label
WAN    *:eb   label +1
2.4 GHz  *:ea   label
The label MAC address found in u-boot at 0x1fc00

Installation:
Upload openwrt-ath79-generic-tplink_tl-wr841hp-v3-squashfs-factory.bin
from stock firmware webgui.
Maybe we need rename to shorten file name due to stock webgui error.

Revert back to stock firmware instructions:
- set your PC to static IP address 192.168.0.66 netmask 255.255.255.0
- download stock firmware from Tp-link website
- put it in the root directory of tftp server software
- rename it to wr841hpv3_tp_recovery.bin
- power on while pressing Reset button until any Led is lighting up
- wait for the router to reboot. done

Forum support topic:
https://forum.openwrt.org/t/support-for-tp-link-tl-wr841hp-v3-router

Signed-off-by: teqlee <congquynh284@yahoo.com>

teqlee committed:
ath79: add support for TP-Link TL-WR841HP v3

Specifications:
- QCA9533 SoC, 8 MB nor flash, 64 MB DDR2 RAM
- 4x Ethernet LAN 10/100, 1x Ethernet WAN 10/100
- 1x WAN, LAN, Wifi, PWR, WPS, RE Leds
- Reset, Wifi on/off, WPS, RE buttons
- Serial UART at J4 onboard: 3.3v GND RX TX, 1152008N1

Label MAC addresses based on vendor firmware:
LAN    *:ea    label
WAN    *:eb   label +1
2.4 GHz  *:ea   label
The label MAC address in found in u-boot 0x1fc00

Installation:
Upload openwrt-ath79-generic-tplink_tl-wr841hp-v3-squashfs-factory.bin
from stock firmware webgui.
Maybe we need rename to shorten file name due to stock webgui error.

Revert back to stock firmware instructions:
- set your PC to static IP address 192.168.0.66 netmask 255.255.255.0
- download stock firmware from Tp-link website
- put it in the root directory of tftp server software
- rename it to wr841hpv3_tp_recovery.bin
- power on while pressing Reset button until any Led is lighting up
- wait for the router to reboot. done

Forum support topic:
https://forum.openwrt.org/t/support-for-tp-link-tl-wr841hp-v3-router

Signed-off-by: teqlee <congquynh284@yahoo.com>

adschm commented:
> Signed-off-by: teqlee ...

Please use your full/real name for Signed-off-by.

adschm commented on target/linux/ath79/dts/qca9533_tplink_tl-wr841hp-v3.dts:
```diff
@@ -0,0 +1,147 @@
+		};
+
+		led_pwr: pwr {
+			label = "green:pwr";
```
Why now "power" instead of "pwr" ...

CodeFetch commented:
@teqlee Can you make a photo of the label or something? Where are they selling a WR841HP? I've never heard of such a device...

adschm commented:
Google gave me that:
https://www.tp-link.com/au/support/download/tl-wr841hp/v3/

Didn't find a source to verify the specs yet, though (without looking too much) ...

teqlee commented:
You can see the picture here https://deviwiki.com/wiki/TP-LINK_TL-WR841HP_v3.x

teqlee committed:
ath79: add support for TP-Link TL-WR841HP v3

Specifications:
- QCA9533 SoC, 8 MB nor flash, 64 MB DDR2 RAM
- 2x2 9dBi antenna, wifi 2.4Ghz 300Mbps
- 4x Ethernet LAN 10/100, 1x Ethernet WAN 10/100
- 1x WAN, LAN, Wifi, PWR, WPS, RE Leds
- Reset, Wifi on/off, WPS, RE buttons
- Serial UART at J4 onboard: 3.3v GND RX TX, 1152008N1

Label MAC addresses based on vendor firmware:
LAN    *:ea    label
WAN    *:eb   label +1
2.4 GHz  *:ea   label
The label MAC address in found in u-boot 0x1fc00

Installation:
Upload openwrt-ath79-generic-tplink_tl-wr841hp-v3-squashfs-factory.bin
from stock firmware webgui.
Maybe we need rename to shorten file name due to stock webgui error.

Revert back to stock firmware instructions:
- set your PC to static IP address 192.168.0.66 netmask 255.255.255.0
- download stock firmware from Tp-link website
- put it in the root directory of tftp server software
- rename it to wr841hpv3_tp_recovery.bin
- power on while pressing Reset button until any Led is lighting up
- wait for the router to reboot. done

Forum support topic:
https://forum.openwrt.org/t/support-for-tp-link-tl-wr841hp-v3-router

Signed-off-by: Andy Lee <congquynh284@yahoo.com>

CodeFetch commented:
@teqlee  Cool device... There are no external PAs on the board. I guess it's an ordinary WR841N v12 but with 12 dBi antennas.

Can you please provide me with a dump of the U-Boot and the ART partition? 
I think the command for dumping ART is:
`dd if=/dev/mtd4 of=/tmp/artdump.bin`
and for U-Boot it is:
`dd if=/dev/mtd0 of=/tmp/artdump.bin`

I've got a repository for collecting and analyzing ath9k ART data:
https://github.com/CodeFetch/art-collection

And a fork of pepe2k's U-Boot mod where I'd like to add this device:
https://github.com/CodeFetch/router-u-boot

teqlee commented:
> 
> @teqlee Cool device... There are no external PAs on the board. I guess it's an ordinary WR841N v12 but with 12 dBi antennas.
> 
> Can you please provide me with a dump of the U-Boot and the ART partition?
>
It has a power amp circuit, do you see it in the picture (SE2576L)? I think this is high power version of wr841n v9/v10, but there is difference in RAM & flash capacity.
Please visit the firmware download link i posted on the support forum, i have uploaded the uboot/art dump you requested.

CodeFetch commented:
@teqlee Very interesting... But that would mean that this device can do more than 10 Watt EIRP :grin:... Is this legal in Australia?

teqlee commented:
@CodeFetch I don't know, i'm not live in Australia. But the OEM has limited Tx power to 24 dBi in the eeprom.

CodeFetch commented on target/linux/ath79/dts/qca9533_tplink_tl-wr841hp-v3.dts:
```diff
@@ -0,0 +1,147 @@
+		led_power: power {
+			label = "green:power";
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+			default-state = "on";
```
Why do you set default-state to on?

teqlee replied:
because factory firmware is default-state on also.

adschm replied:
Please _keep_ the default-state = on

CodeFetch replied:
Why do we set them off as a default in DTS then?

CodeFetch replied:
@adschm I think this whole inconsistent default state on/off does not make sense...

adschm replied:
I do not understand the question. The default for _any_ LED is "off", since it would not make sense to have _all_ LEDs enabled by default. And then we use the default-state keyword to define for which LEDs we would have a different behavior.

adschm replied:
And please open a separate thread somewhere for this discussion, so we do not confuse submitters with off-topic taste discussions. default-state = on is correct here.

adschm commented on target/linux/ath79/dts/qca9533_tplink_tl-wr841hp-v3.dts:
```diff
@@ -0,0 +1,147 @@
+
+		re {
+			label = "RE button";
+			linux,input-type = <EV_SW>;
```
Based on the manual, this seems to be a button, and not a switch?

teqlee replied:
yep, a button, but how do I define it ?

adschm replied:
Just remove the input-type line, button is "default" (like for all other buttons you define). The only reason to include EV_SW for a button would be if the button has two "permanent" states, i.e. like some hardware power buttons.

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -190,7 +190,8 @@ ath79_setup_interfaces()
+	tplink,tl-wr841hp-v3)
 		ucidef_set_interface_wan "eth1"
 		ucidef_add_switch "switch0" \
 			"0@eth0" "1:lan" "2:lan" "3:lan" "4:lan"
```
Normal wr841 has inverted port labels:
`"0@eth0" "1:lan:4" "2:lan:3" "3:lan:2" "4:lan:1"`

Are yours correct?

teqlee replied:
Yep, i'm sure it's correct. Plugin cable to ethernet port 1 on router, switch status on luci gui show Lan 1.

adschm replied:
Perfect, thanks.

adschm commented on target/linux/ath79/dts/qca9533_tplink_tl-wr841hp-v3.dts:
```diff
@@ -0,0 +1,147 @@
+			default-state = "on";
+		};
+
+		wan {
```
wan_green

teqlee replied:
ok, noted.

teqlee replied:
@adschm I changed everything as you requested, when will it be accept to merge ?

teqlee committed:
ath79: add support for TP-Link TL-WR841HP v3

Specifications:
- QCA9533 SoC, 8 MB nor flash, 64 MB DDR2 RAM
- 2x2 9dBi antenna, wifi 2.4Ghz 300Mbps
- 4x Ethernet LAN 10/100, 1x Ethernet WAN 10/100
- 1x WAN, LAN, Wifi, PWR, WPS, RE Leds
- Reset, Wifi on/off, WPS, RE buttons
- Serial UART at J4 onboard: 3.3v GND RX TX, 1152008N1

Label MAC addresses based on vendor firmware:
LAN    *:ea    label
WAN    *:eb   label +1
2.4 GHz  *:ea   label
The label MAC address in found in u-boot 0x1fc00

Installation:
Upload openwrt-ath79-generic-tplink_tl-wr841hp-v3-squashfs-factory.bin
from stock firmware webgui.
Maybe we need rename to shorten file name due to stock webgui error.

Revert back to stock firmware instructions:
- set your PC to static IP address 192.168.0.66 netmask 255.255.255.0
- download stock firmware from Tp-link website
- put it in the root directory of tftp server software
- rename it to wr841hpv3_tp_recovery.bin
- power on while pressing Reset button until any Led is lighting up
- wait for the router to reboot. done

Forum support topic:
https://forum.openwrt.org/t/support-for-tp-link-tl-wr841hp-v3-router

Signed-off-by: Andy Lee <congquynh284@yahoo.com>

adschm commented:
Thanks, I will merge a squashed version during the weekend.
